### ISSUE TITLE
Streaming may hold list ofsstables to read from for a long duration

### ISSUE THREAD

**Issue by slivne** [scylladb/scylla#2860]
Readers created by streaming may hold onto compacted away sstables for a long duration.

Allow replacing the compacted away sstables by their replacements.

> Suppose an sstable_list sl1 has sstables s1, s2, s3, .. sn. Suppose a compaction rewrites sc1, sc2, ... scm as sr1, sr2, ... srk, If all of sc1, ... scm are present in sl1, then we can replace them with sr1, sr2, ... srk.
> 
> 
> More relaxed,
> 
> 
> If sc1, sc2, ... scm all appeared in the sstable_set from which sl1 was derived, then they can be replaced by s1, s2, ... srk, or a subset thereof (to account for token-range limits in LCS or clustering-key constraints with DTCS/TWCS.
> 
> 
> (In plain language if the compaction did not include a newer sstable than the sstable_set that was used as a basis for that sstable_list had, then we can replace them; we can even have sstable_list (or perhaps snapshot_source) do it transparently).
> 

additional discussion on the internal mail thread "Why do we need to hold sstables during queued reads?"

**Comment by denesb** [comment]
@raphaelsc is this issue still valid?

**Comment by raphaelsc** [comment]
@denesb Yes. We don't propagate yet sstable replacement triggered by compaction. This will overlap with my work on sstable run which needs to get reference of compacted ssts. This issue can end up being fixed as a result.

**Comment by raphaelsc** [comment]
this is specially interesting for incremental compaction
